### PR TITLE
[BEAM-27] Add analysis of Timer parameters for DoFn.ProcessElement

### PR DESCRIPTION
**Pull request by kennknowles**
Be sure to do all of the following to help us incorporate your contribution
quickly and easily:

 - [x] Make sure the PR title is formatted like:
   `[BEAM-<Jira issue #>] Description of pull request`
 - [x] Make sure tests pass via `mvn clean verify`. (Even better, enable
       Travis-CI on your fork and ensure the whole test matrix passes).
 - [x] Replace `<Jira issue #>` in the title with the actual Jira issue
       number, if there is one.
 - [x] If this contribution is large, please file an Apache
       [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.txt).

---

This PR adds `Timer` parameters. It includes #1102 and #1130. The final commit is the only one that is new, and can be reviewed independently.

R: @jkff and @bjchambers

### CODE REVIEW

File: sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/reflect/DoFnSignatures.java [hunk 1]
```diff
@@ -505,6 +513,56 @@ private static void verifyUnsplittableMethods(ErrorReporter errors, DoFnSignatur
...
+            fnClass.getRawType().getName(),
+            params[i],
+            i,
+            DoFn.TimerId.class.getSimpleName(),
+            id);
+
+        TimerDeclaration timerDecl = timerDeclarations.get(id);
+        errors.checkArgument(
+            timerDecl != null,
+            "%s parameter of type %s at index %s references undeclared StateId \"%s\"",
```
Comment (jkff) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/reflect/DoFnSignatures.java]: undeclared timer, you mean?

File: sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/reflect/DoFnSignatures.java [hunk 2]
```diff
@@ -505,6 +513,56 @@ private static void verifyUnsplittableMethods(ErrorReporter errors, DoFnSignatur
...
+            fnClass.getRawType().getName(),
+            params[i],
+            i,
+            id);
+
+        errors.checkArgument(
+            timerDecl.field().getDeclaringClass().equals(m.getDeclaringClass()),
+            "Method %s has State parameter at index %s for state %s"
```
Comment (jkff) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/reflect/DoFnSignatures.java]: Another copy-paste error. Please check for more, or consider finding a way to deduplicate.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/reflect/DoFnSignatures.java [hunk 3]
```diff
@@ -505,6 +514,58 @@ private static void verifyUnsplittableMethods(ErrorReporter errors, DoFnSignatur
             formatType(paramT),
             formatType(expectedOutputReceiverT));
         extraParameters.add(DoFnSignature.Parameter.outputReceiver());
+      } else if (Timer.class.equals(rawType)) {
```
Comment (bjchambers) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/reflect/DoFnSignatures.java]: It seems like the code in this block could be extracted into a helper method pretty easily?

Comment (kennknowles) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/reflect/DoFnSignatures.java]: I think I'm going to directly follow up with splitting this out.

**Pull request merged**